### ISSUE TITLE
Shared memory interface

### ISSUE THREAD

**Issue by jklmnn** [Componolit/gneiss#110]
- [x] Define shared memory interface with client semantics
- [x] Implement for Linux using `memfd`

**Comment by jklmnn** [comment]
The shared memory interface simply provides a single memory region to two components. It should be generic in the sense that it takes a byte and byte array type. The management of the memory is part of the application or higher level interface.

**Comment by jklmnn** [comment]
The shared memory interface will be a client only interface that can access a named memory (which can also be a file, and be accessed by other components). The server will always be the platform. It will also replace the ROM session as it can be configured to be read only. The memory will be accessed via a callback that, depending on the configuration will provide an `in` or `in out` parameter.
```Ada
generic
   type Element is (<>);
   type Index is range <>;
   type Buffer is array (Index range <>) of Element;
package Gneiss.Memory with
   SPARK_Mode
is

   type Access_Mode is (Read_Only, Read_Write);

end Gneiss.Memory;

generic
   with procedure Update (Session : in out Client_Session;
                          Data    :        Buffer);
   with procedure Update (Session : in out Client_Session;
                          Data    : in out Buffer);
package Gneiss.Memory.Client with
   SPARK_Mode
is

   procedure Initialize (Session : in out Client_Session;
                         Cap     :        Gneiss.Capability;
                         Label   :        String;
                         Mode    :        Access_Mode;
                         Idx     :        Gneiss.Session_Index := 1);

   function Size (Session : Client_Session) return Index;

   procedure Update (Session : in out Client_Session);

end Gneiss.Memory.Client;
```

On Linux a `Memory` session will not point to a server but a resource:
```XML
<config>
    <resource name="res" type="Memory"/>
    <resource name="config" type="File" location="/etc/config"/>
    <component name="test" file="libcomponent_test.so">
        <service name="Memory" resource="res" writable="True"/>
        <service name="Memory" resource="config" writable="False"/>
    </component>
</config>
```

**Comment by jklmnn** [comment]
`memfd_create` can only open writable file descriptors. Also sealing does apply to the whole file and not only to the file descriptor. The concept is to seal the opened shared memory region with `F_SEAL_SHRINK`, `F_SEAL_GROW` and `F_SEAL_SEAL` to fix the size and the seals. The client file descriptor will be obtained by opening `/proc/self/fd/<fd>` with the desired flags.

**Comment by jklmnn** [comment]
Fixed by #141.
